add checks for OpenCL C versions

bashbaug opened the pull request:
fixes #897 
fixes #899 

This PR adds a new test to the `compiler` suite.  There are four sub-tests:

1. For the value returned by `CL_DEVICE_OPENCL_C_VERSION`, first check that it meets any minimum API requirements.  If it does, verify that a program can be built using build options for the reported version and all previous versions.  This sub-test is not OpenCL 3.0-specific and can run on devices that support earlier OpenCL versions.

2. For the versions returned by `CL_DEVICE_OPENCL_C_ALL_VERSIONS`, verify that a program can be built using build options for each reported version.  Also, check that any required API versions are supported.  Since `CL_DEVICE_OPENCL_C_ALL_VERSIONS` is a new query in OpenCL 3.0, this sub-test is OpenCL 3.0-specific.

3. For some values returned by `CL_DEVICE_OPENCL_C_VERSION` (currently only OpenCL C 2.0), verify that the compiler reports the required features in `CL_DEVICE_OPENCL_C_FEATURES`.  Since `CL_DEVICE_OPENCL_C_FEATURES` is a new query in OpenCL 3.0 this sub-test is OpenCL 3.0-specific.

4. Verify that each of the versions equal to or smaller than `CL_DEVICE_OPENCL_C_VERSION` are also returned in  `CL_DEVICE_OPENCL_C_ALL_VERSIONS`.  Again, since `CL_DEVICE_OPENCL_C_ALL_VERSIONS` is a new query in OpenCL 3.0, this sub-test is OpenCL 3.0-specific.

bcalidas commented:
Qualcomm has tested this PR and it passed for us.

FranklandJack reviewed (dismissed):
Codeplay have tested this locally and it is working for us with a 3.0 implementation that does not support `-cl-std=CL2.0`.

bashbaug commented on test_common/harness/featureHelpers.cpp:
```diff
@@ -0,0 +1,77 @@
+
+int get_device_cl_c_features(cl_device_id device, OpenCLCFeatures* features)
+{
+    assert(features != NULL);
```
Good point - fixed.

We could also just return the structure by value if we don't care about returning a possible error value, since this function shouldn't be called frequently.

alycm reviewed (approved):
LGTM, I did make some comments but if it comes down to it none of them are blockers.

bashbaug commented on test_common/harness/featureHelpers.h:
```diff
@@ -0,0 +1,43 @@
+    bool __opencl_c_program_scope_global_variables;
+    bool __opencl_c_read_write_images;
+    bool __opencl_c_subgroups;
+    bool __opencl_c_work_group_collective_functions;
```
Fixed by adding a "supports" prefix, so:

```c
    bool supports__opencl_c_program_scope_global_variables;
```

etc.

bashbaug commented on test_conformance/compiler/test_opencl_c_versions.cpp:
```diff
@@ -0,0 +1,315 @@
+}
+
+// This sub-test checks that all required OpenCL C versions are present for a
+// specific C_DEVICE_OPENCL_C_VERSION.
```
Fixed!

bashbaug commented on test_conformance/compiler/test_opencl_c_versions.cpp:
```diff
@@ -0,0 +1,315 @@
+        if (features.__opencl_c_images)
+        {
+            has_all_OpenCL_C_20_features = has_all_OpenCL_C_20_features
+                && features.__opencl_c_read_write_images;
```
Yes, I believe so, good catch.

bashbaug commented:
@alycm I think I have addressed all of your review comments, thanks!

For the other reviewers, the only functional change was to require `__opencl_c_3d_image_writes` if the device supports OpenCL C 2.0.  All of the other changes were cosmetic.

alycm reviewed (approved):
LGTM.

alycm commented:
This has been reviewed twice (though by the same company) and a third person has at least run the tests.  I'll merge this.
